samanthas_telegram_bot/api_clients/base/base_api_client.py

```diff
@@ -0,0 +1,274 @@
+                response = await cls._make_one_request(
+                    method=method,
+                    url=url,
+                    headers=headers,
+                    data=data,
+                    json_data=json_data,
+                    params=params,
+                )
+            except (NotImplementedError, httpx.TransportError) as err:
+                # TODO logic can be split here, maybe let user try again in case of httpx error.
+                raise BaseApiClientError(
+                    f"Failed to send {method.upper()} request to {url=}"
+                ) from err
+
+            if response.status_code not in (
+                httpx.codes.INTERNAL_SERVER_ERROR,
+                httpx.codes.BAD_GATEWAY,
+                httpx.codes.SERVICE_UNAVAILABLE,
+                httpx.codes.GATEWAY_TIMEOUT,
+            ):
+                return response
+
+            attempts += 1
+            timeout *= 2
+
+            await cls._log_retry(
+                update=update,
+                context=context,
+                url=url,
+                method=method,
+                status_code=response.status_code,
+                attempts=attempts,
+                timeout=timeout,
+            )
+            await asyncio.sleep(timeout)
+
+    @staticmethod
+    async def _make_one_request(
+        method: HttpMethod,
+        url: str,
+        headers: dict[str, str] | None = None,
+        data: DataDict | None = None,
+        json_data: DataDict | None = None,
+        params: DataDict | None = None,
+    ) -> Response:
+        async with httpx.AsyncClient() as client:
+            if method == HttpMethod.GET:
+                response = await client.get(url, headers=headers, params=params)
+            elif method == HttpMethod.POST:
+                response = await client.post(
+                    url, headers=headers, params=params, data=data, json=json_data
+                )
+            else:
+                raise NotImplementedError(f"{method=} not supported")
+
+        logger.debug(
+            f"Sent {method.upper()} request to {url=} with {data=}. {response.status_code=}."
+        )
+        return response
+
+    @staticmethod
+    async def _log_retry(
+        update: Update,
+        context: CUSTOM_CONTEXT_TYPES,
+        url: str,
+        method: HttpMethod,
+        status_code: int,
+        timeout: int,
+        attempts: int,
+    ) -> None:
+        if attempts == 1:
+            await logs(
+                bot=context.bot,
+                update=update,
+                text=(
+                    f"Failed to reach {url=} with {method=} ({status_code=}). "
+                    f"Will try {MAX_ATTEMPTS_TO_GET_DATA_FROM_API - attempts} times before failing"
+                ),
+                level=LoggingLevel.WARNING,
+                needs_to_notify_admin_group=True,
+            )
+            return
+
+        if attempts == MAX_ATTEMPTS_TO_GET_DATA_FROM_API:
```

Comment:
I think logging function should only log something, not check attempts amount, this should be moved to `_make_request_with_retries`

Reply:
Yep, it was actually a duplicate of what was already in `_make_request_with_retries()`. Removed.